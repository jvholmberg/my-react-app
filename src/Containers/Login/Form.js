import React from 'react';
import { Formik, Form, Field } from 'formik';
import Yup from 'yup';

import { TextInput, Button } from '../../Components';

const validationSchema = (values) => Yup.object().shape({
  email: Yup.string()
    .email('Invalid email address')
    .required('Email is required!'),
  password: Yup.string()
    .min(8, 'Password must be atleast 8 characters')
    .required('Password is required'),
});

const handleSubmit  = (values, actions) => {
  setTimeout(() => {
    alert(JSON.stringify(values, null, 2));
    actions.setSubmitting(false);
  }, 1000);
};

const renderErrors = (errors, touched) => {
  // Single error
  if (touched && typeof errors === 'string') {
    return <div className='LoginForm-Error'>* {errors}</div>;
  }
  // Multiple errors
  if (touched && errors) {
    return errors.map((err) => <div className='LoginForm-Error'>* {err}</div>);
  }
  // No errors
  return <div></div>;
};

const render = (formProps) => (
  <Form>
    <Field
      name="email"
      render={({ field, form: { touched, errors } }) => (
        <TextInput label='Email' required touched={touched} errors={errors} {...field} />
      )} />
    <Field
      name="password"
      render={({ field, form: { touched, errors } }) => (
        <TextInput label='Password' required password touched={touched} errors={errors} {...field} />
      )} />
    <Button primary medium type='submit' disabled={formProps.isSubmitting}>
      Log in
    </Button>
    <Button secondary medium onClick={formProps.handleReset}>
      Reset
    </Button>
    {renderErrors(formProps.errors.email, formProps.touched.email)}
    {renderErrors(formProps.errors.password, formProps.touched.password)}
  </Form>
);

const LoginForm = () => (
  <Formik
    initialValues={{ email: '', password: '' }}
    validationSchema={validationSchema}
    onSubmit={handleSubmit}
    render={render}
  />
);

export default LoginForm;
