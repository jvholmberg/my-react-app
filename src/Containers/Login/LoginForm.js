import React from 'react';
import { Field, reduxForm } from 'redux-form'
import { TextInput, Button } from '../../Components';

const EmailInput= () =>
  <TextInput label='Email' primary required />;

const PasswordInput= () =>
  <TextInput label='Password' primary password required />;

const validate = (values) => {
  const errors = {};

  if (!values.email) {
     errors.email = 'Required'
   } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
     errors.email = 'Invalid email address'
   }

   return errors;
}

const warn = (values) => {
  console.log(values);
};

const LoginForm = ({ handleSubmit, pristine, reset, submitting }) => (
  <form onSubmit={handleSubmit}>
    <Field name='email' component={EmailInput} />
    <Field name='password' component={PasswordInput} />
    <Button
      primary
      medium
      type='submit'
      disabled={submitting}>
        Login
    </Button>
    <Button
      secondary
      medium
      onClick={reset}
      disabled={pristine || submitting}>
        Reset
    </Button>
  </form>
);

export default reduxForm({
  form: 'loginForm',
  validate,
  warn,
})(LoginForm);
