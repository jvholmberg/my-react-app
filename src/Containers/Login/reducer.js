
const SUBMIT_LOGIN_FORM = 'SUBMIT_LOGIN_FORM';

export const submitLoginForm = (values) =>
  ({ type: SUBMIT_LOGIN_FORM, values });

export default function (state = {}, action) {
  switch (action.type) {
    case SUBMIT_LOGIN_FORM:
      return action.values;
    default:
      return state;
  }
}
