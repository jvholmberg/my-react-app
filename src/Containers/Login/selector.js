import _ from 'lodash';

export const getLoginFormValues = (state) =>
  _.get(state, 'loginForm.values', false);
