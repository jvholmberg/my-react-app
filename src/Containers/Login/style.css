.LoginForm-Error {
  padding: 1rem 0 0;
  font-weight: 300;
  color: #ca3a47;
}
