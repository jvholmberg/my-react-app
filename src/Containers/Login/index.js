import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { Container, Grid, Footer } from '../../Components';
import LoginForm from './LoginForm';
import LoginFormReducer, { submitLoginForm } from './reducer';

const mapStateToProps = (state) => {
  return ({});
};

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators({
    submitLoginForm,
  }, dispatch),
});

const Login = ({ actions }) => (
  <Container full>
    <Container regular>
      <Grid col3></Grid>
      <Grid col6>
        <LoginForm onSubmit={(values) => {
          console.log(values);
          actions.submitLoginForm(values);
        }} />
      </Grid>
      <Grid col3></Grid>
    </Container>
    <Footer />
  </Container>
);
export default connect(mapStateToProps, mapDispatchToProps)(Login);
export { LoginFormReducer }
