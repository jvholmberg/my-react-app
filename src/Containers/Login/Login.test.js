import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Login from './';
import LoginForm from './LoginForm';

describe('LoginForm', () => {
  it('Empty matches snapshot', () => {
    const tree = renderer
      .create(<LoginForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('submit works', () => {
    const submitMock = () => Promise.resolve();
    const loginForm = mount(<LoginForm handleSubmit={submitMock} />);
    expect(loginForm.state().isSubmitting).toBe(false);
    loginForm.find('form').simulate('submit', {
        preventDefault() {
          // this gets called because it's sync
        },
      });
    console.log(loginForm.state().isSubmitting);
  });
});
