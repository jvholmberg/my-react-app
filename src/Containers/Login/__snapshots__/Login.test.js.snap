// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Empty matches snapshot 1`] = `
<form
  className="LoginForm-Container"
  onSubmit={[Function]}
>
  <div
    className="TextInput-Container"
  >
    <label
      className="TextInput-Label"
    >
      Email
       
      <span
        className="TextInput-Required"
      >
        *
      </span>
      <input
        className="TextInput Primary"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder={undefined}
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    className="TextInput-Container"
  >
    <label
      className="TextInput-Label"
    >
      Password
       
      <span
        className="TextInput-Required"
      >
        *
      </span>
      <input
        className="TextInput Primary"
        id="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder={undefined}
        type="password"
        value=""
      />
    </label>
  </div>
  <button
    className="Button Primary Button-Size-Medium"
    disabled={false}
    type="submit"
  >
    Login
  </button>
  <button
    className="Button Secondary Button-Size-Medium"
    disabled={true}
    onClick={[Function]}
  >
    Reset
  </button>
  <div />
  <div />
</form>
`;
